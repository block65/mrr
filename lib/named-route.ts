import type { Params, ExtractRouteParams } from './types.js';
import { nullOrigin, urlObjectAssign, urlRhs } from './util.js';

type QueryParams = Record<string, string>;

export interface NamedRoute<P extends string, Q extends QueryParams> {
  path: P;
  build: (x?: {
    params?: ExtractRouteParams<P>;
    query?: Q;
    hash?: string;
    origin?: string;
  }) => string;
}

export function interpolate<P extends string>(
  path: P,
  params: ExtractRouteParams<P> | Params = {},
): string {
  return path.replace(/\/:(\w+)[?+*]?/g, (_match, token: keyof typeof params) =>
    params[token] ? `/${params[token]}` : '',
  );
}

export function namedRoute<T extends string, Q extends QueryParams = never>(
  pathname: T,
): NamedRoute<T, Q> {
  return {
    path: pathname,
    build({ params, query, origin, ...rest } = {}) {
      const interpolatedPathname = params
        ? interpolate(pathname, params)
        : pathname;

      const searchParams = new URLSearchParams(query);
      searchParams.sort();

      const newUrl = urlObjectAssign(new URL(origin || nullOrigin), {
        pathname: interpolatedPathname,
        search: searchParams.toString(),
        ...rest,
      });

      return newUrl.origin === nullOrigin.origin
        ? urlRhs(newUrl)
        : newUrl.toString();
    },
  };
}
