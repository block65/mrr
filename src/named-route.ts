export { namedRoute } from '../lib/named-route.js';

export type {
  NamedRoute,
  NamedRouteWithParams,
  NamedRouteWithoutParams,
} from '../lib/named-route.js';
