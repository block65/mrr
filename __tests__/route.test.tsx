import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import { useCallback, useEffect, type FC } from 'react';
import { expect, test, vi } from 'vitest';
import { namedRoute } from '../lib/named-route.js';
import type { RouteComponentProps } from '../lib/types.js';
import {
  Route,
  Router,
  Routes,
  type SyntheticNavigateEvent,
} from '../src/index.js';

const login = namedRoute('/');

test('custom route', async () => {
  const CustomRoute = <T extends string>(props: RouteComponentProps<T>) => (
    <Route {...props} />
  );

  const { asFragment } = render(
    <Router>
      <Routes>
        <CustomRoute path={login.path}>
          <h1>custom route</h1>
        </CustomRoute>
      </Routes>
    </Router>,
  );

  expect(asFragment()).toMatchSnapshot();
});

test('cancel nav', async () => {
  const Component: FC = () => {
    const onNav = useCallback(async (e: SyntheticNavigateEvent) => {
      e.preventDefault();
    }, []);
    return (
      <Router intercept={onNav} useNavApi={false}>
        <Routes>
          <Route path={login.path}>
            <h1>custom route</h1>
          </Route>
        </Routes>
      </Router>
    );
  };

  const { asFragment } = render(<Component />);

  expect(asFragment()).toMatchSnapshot();
});

test('routes with components/children/paths/no paths', async () => {
  const HelloComponent = () => <h1>hello</h1>;

  const { asFragment } = render(
    <Router useNavApi={false}>
      <Routes>
        <Route path={login.path} component={HelloComponent} />
        <Route path={login.path}>
          <HelloComponent />
        </Route>
        <Route component={HelloComponent} />
        <Route>
          <HelloComponent />
        </Route>
      </Routes>
    </Router>,
  );

  expect(asFragment()).toMatchSnapshot();
});

test('effects inside route components or children dont fire', async () => {
  const effectFn = vi.fn();

  const HelloComponent: FC = () => {
    useEffect(() => {
      effectFn();
    }, []);
    return <h1>hello</h1>;
  };

  render(
    <Router useNavApi={false}>
      <Routes>
        <Route path="/random" component={HelloComponent} />
        <Route path="random2">
          <HelloComponent />
        </Route>
      </Routes>
    </Router>,
  );

  expect(effectFn).not.toHaveBeenCalled();
});
