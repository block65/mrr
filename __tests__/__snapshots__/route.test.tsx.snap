// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom route 1`] = `
<DocumentFragment>
  <h1>
    custom route
  </h1>
</DocumentFragment>
`;
